Fix/accept product option metadata

fogine opened the pull request:
fixes #11424

olivermrbl reviewed (commented):
appreciate the contribution! 

could I get you to add tests for this? I asumme you want to update metadata using the Rest API, so you can add the tests in `integration-tests/http/...`

fogine commented:
> appreciate the contribution!
> 
> could I get you to add tests for this? I asumme you want to update metadata using the Rest API, so you can add the tests in `integration-tests/http/...`

@olivermrbl  Thank for looking into this.  
My intention is to update/create product options via medusa's `batchProductsWorkflow` internally.
I'll look into adding integration test for this. Could you point me to the Rest API endpoint that allows to set product options?

fogine committed:
fix/importProductsWorkflow and create/update/upsert product option DTO types - allow to set product option metadata and product option value metadata

fogine committed:
Merge branch 'develop' into fix/accept-product-option-metadata

fogine committed:
fix: tests

fogine committed:
feat: add changeset

fogine committed:
fix: integration tests

fogine committed:
fix: revert mistakenly commited debug state of test file

fogine committed:
fix: variant option validation schema and product variant options csv import

fogine committed:
fix: integration tests and docs

fogine committed:
feat: replace changeset

fogine committed:
fix: product option validation schema

fogine committed:
feat: allow to set product option metadata via admin API endpoints

fogine committed:
feat: add test case for POST and GET admin API product endpoint that verifies product option metadata insertion and retrieval

fogine committed:
feat: add more test cases

fogine commented:
I have fixed all integration and unit tests and added test case verifying the user can create product option and product option value metadata and fetch them using POST and GET admin (and store) API endpoints respectively.

At the end of the day, it took significantly more time to implement than I expected. Mainly due to frequent use of `any` types throughout the project. IMHO getting rid off excessive use of ts `any` types should be on top of the priority list. I believe getting rid off them would significantly decrease volume of bugs and inconsistencies introduced during the future development.

There is breaking change in the admin create product API and relevant workflows. Product options now accept value container object with `value` and `metadata` properties:   

```typescript
    {
         //...
          options: [
            {
              title: "size",
              metadata: {arbitrary: 'value'},
              values: [
                {
                  value: "large",
                  metadata: {arbitrary: 'data'}
                }
              ]
            }],
    }

```

instead of original `values` being array of strings without metadata:


```typescript
    {
         //...
          options: [
            {
              title: "size",
              values: ['large']
            }],
    }
```

willbouch commented:
@olivermrbl how do you feel about the breaking change? I'm a bit iffy on that one

willbouch commented:
Hi @fogine! I spoke with the team and considering that we plan on reworking the option metadata in the near future (and therefore will make changes to the API), we decided not to go forward with a breaking change for now unfortunately. Will be closing the PR but feel free to re-open if you can introduce the change without introducing a breaking change
